<h1 id="introduction">Introduction</h1>
<p>This vignette describes the <em>methQTL</em> R-package (https://github.com/MPIIComputationalEpigenetics/methQTL-package) available from GitHub. The package uses DNA methylation data obtained using the Illumina BeadArrays, and genotyping data from Illumina genotyping microarrays or whole genome sequencing to compute methylation quantitative trait loci (methQTL). The package provides mutliple flavors of linear modeling strategies to compute <em>methQTL</em> as statistically significant interactions between single nucleotide polymorphisms (SNPs) and changes in the DNA methylation state of individual CpGs. DNA methylation values at single CpGs are first summarized into correlation blocks, and a representative of this correlation block is used for methQTL calling.</p>
<h1 id="installation">Installation</h1>
<p>The package can be directly installed from GitHub, after installing the <em>devtools</em> package.</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb1-1" data-line-number="1"><span class="cf">if</span>(<span class="op">!</span><span class="kw">requireNamespace</span>(<span class="st">&quot;devtools&quot;</span>)) <span class="kw">install.packages</span>(<span class="st">&quot;devtools&quot;</span>)</a></code></pre></div>
<pre><code>## Loading required namespace: devtools</code></pre>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb3-1" data-line-number="1"><span class="cf">if</span>(<span class="op">!</span><span class="kw">requireNamespace</span>(<span class="st">&quot;methQTL&quot;</span>)){</a>
<a class="sourceLine" id="cb3-2" data-line-number="2">  devtools<span class="op">::</span><span class="kw">install_github</span>(<span class="st">&quot;MPIIComputationalEpigenetics/methQTL-package&quot;</span>)</a>
<a class="sourceLine" id="cb3-3" data-line-number="3">}  </a></code></pre></div>
<pre><code>## Loading required namespace: methQTL
## Setting options(&#39;download.file.method.GEOquery&#39;=&#39;auto&#39;)
## Setting options(&#39;GEOquery.inmemory.gpl&#39;=FALSE)
## Warning: no function found corresponding to methods exports from &#39;RnBeads&#39; for:
## &#39;samples&#39;</code></pre>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb5-1" data-line-number="1"><span class="kw">suppressPackageStartupMessages</span>(<span class="kw">library</span>(methQTL))</a></code></pre></div>
<h1 id="input-data">Input data</h1>
<p>The <em>methQTL</em> package requires two types of data as input: DNA methylation data obtained using the Illumina Infinium BeadArrays and genotyping data obtained using genotyping microarrays or whole genome sequencing.</p>
<h2 id="dna-methylation-data-microarrays">DNA methylation data (microarrays)</h2>
<p>The package utilizes the widely used <a href="http://rnbeads.org/"><em>RnBeads</em></a> software package for DNA methylation data import. It supports the various input options available in <em>RnBeads</em>, including a direct download from the Gene Expression Omnibus (GEO). For further options, we refer to the <a href="http://bioconductor.org/packages/release/bioc/vignettes/RnBeads/inst/doc/RnBeads.pdf">RnBeads vignette</a> and documentation. In addition to the raw methylation data, a sample annotation sheet specifying the samples to be analyzed needs to be provided. The sheet contains a line for each sample and looks as follows:</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode bash"><code class="sourceCode bash"><a class="sourceLine" id="cb6-1" data-line-number="1"><span class="ex">SampleID</span>,age,sex,barcode</a>
<a class="sourceLine" id="cb6-2" data-line-number="2"><span class="ex">Sample_1</span>,14,f,209054857842_R01C01</a>
<a class="sourceLine" id="cb6-3" data-line-number="3"><span class="ex">Sample_2</span>,42,f,209054857842_R02C01</a>
<a class="sourceLine" id="cb6-4" data-line-number="4"><span class="ex">Sample_3</span>,45,m,209054857842_R03C01</a></code></pre></div>
<p>For further details on the import process, we refer to the <a href="http://bioconductor.org/packages/release/bioc/vignettes/RnBeads/inst/doc/RnBeads.pdf">RnBeads vignette</a>. Most importantly, analysis options need to be specified for the import and preprocessing modules of <em>RnBeads</em>. <em>methQTL</em> provides a default setting, which is available in <em>extdata/rnbeads_options.xml</em>. You can use this file as a template for your own setting and then specify it to the methQTL package:</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb7-1" data-line-number="1">opts &lt;-<span class="st"> </span><span class="kw">rnb.xml2options</span>(<span class="kw">system.file</span>(<span class="st">&quot;extdata/rnbeads_options.xml&quot;</span>,<span class="dt">package=</span><span class="st">&quot;methQTL&quot;</span>))</a>
<a class="sourceLine" id="cb7-2" data-line-number="2"><span class="kw">rnb.options</span>(<span class="dt">identifiers.column=</span><span class="st">&quot;ind_IPC&quot;</span>)</a>
<a class="sourceLine" id="cb7-3" data-line-number="3">xml.fi &lt;-<span class="st"> </span><span class="kw">file.path</span>(<span class="kw">getwd</span>(),<span class="st">&quot;rnbeads_options.xml&quot;</span>)</a>
<a class="sourceLine" id="cb7-4" data-line-number="4"><span class="kw">cat</span>(<span class="kw">rnb.options2xml</span>(),<span class="dt">file=</span>xml.fi)</a>
<a class="sourceLine" id="cb7-5" data-line-number="5"><span class="kw">qtl.setOption</span>(<span class="dt">rnbeads.options =</span> xml.fi)</a></code></pre></div>
<p>To redefine the correlation blocks, we allow including additional information such as genome-wide segmentation of the methylation landscape (see option <code>use.segmentation</code> and function <code>qtl.run.segmentation</code>), and also function annotation according to the Ensembl regulatory build <span class="citation" data-cites="Zerbino2015">(Zerbino et al. 2015)</span>.</p>
<h2 id="genotyping-data">Genotyping data</h2>
<h3 id="plink-files">PLINK files</h3>
<p>The package supports data that has already been processed by <em>plink</em> and is available either in the form of binary <em>.bed</em>, <em>.bim</em> and <em>.fam</em> files, as <em>.ped</em> and <em>.map</em>, as variant calling files (<em>.vcf</em>), or as imputed files in the dosage format (<em>.dos</em>). For further processing, we use the command line tool <em>plink</em>, which comes with this package, but is only applicable on Linux systems. For Windows and MacOS users, please install the <em>plink</em> tool from <a href="https://www.cog-genomics.org/plink/1.9/">here</a> and specify it using the option <code>plink.path</code>. The sample identifier specified earlier also needs to match the sample IDs of the genotype calls.</p>
<h3 id="idat-files">IDAT files</h3>
<p>The package also support raw IDAT files and uses the <a href="https://www.bioconductor.org/packages/release/bioc/html/crlmm.html">CRLMM</a> R-package, together with PLINK to perform genotype calling and data import. The package requires a single sample annotation sheet in the format described in the <a href="DNA%20methylation%20data">DNA methylation data (microarray)</a> section. In addition to the column names specified above, a column named <em>GenoSentrixPosition</em> has to be added, which specifies the IDAT file IDs.</p>
<div class="sourceCode" id="cb8"><pre class="sourceCode bash"><code class="sourceCode bash"><a class="sourceLine" id="cb8-1" data-line-number="1"><span class="ex">SampleID</span>,age,sex,barcode,GenoSentrixPosition</a>
<a class="sourceLine" id="cb8-2" data-line-number="2"><span class="ex">Sample_1</span>,14,f,209054857842_R01C01,9701756058_R05C01</a>
<a class="sourceLine" id="cb8-3" data-line-number="3"><span class="ex">Sample_2</span>,42,f,209054857842_R02C01,9701756058_R07C01</a>
<a class="sourceLine" id="cb8-4" data-line-number="4"><span class="ex">Sample_3</span>,45,m,209054857842_R03C01,9742011016_R04C01</a></code></pre></div>
<h3 id="imputation">Imputation</h3>
<p>Illumina SNP BeadArray data is typically imputed before further analysis, and the package allows for imputation through the <a href="https://imputationserver.sph.umich.edu/index.html#!">Michigan Imputation Server</a>. In order to be able to perform computation on the server, an account is required. After the account is created, one has to request an API token in the user settings and specify it to the package using the option <code>imputation.user.token</code>. During the imputation process, the package will stall for a while and wait for the job to finish. After the job is done, the package will prompt for entering the password send via e-mail to the user account. The imputation process has to be split according to chromosomes, which is why multiple e-mails will be send to the account, and the imputation process can take up to several days. However, after imputation, the imputed data will be available as PLINK files, such that the imputation has to be performed only once. For preprocessing the data for upload to the imputation server, the package requires the <a href="http://www.htslib.org/doc/bgzip.html">bgzip</a> and <a href="http://www.htslib.org/doc/tabix.html">tabix</a> tools from the <a href="http://www.htslib.org/">htslib</a> package. Also see further options to configure the imputation jobs (see the Michigan Imputation Server <a href="https://imputationserver.readthedocs.io/en/latest/">documentation</a> for further information):</p>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb9-1" data-line-number="1"><span class="kw">qtl.setOption</span>(</a>
<a class="sourceLine" id="cb9-2" data-line-number="2">  <span class="dt">impute.geno.data=</span><span class="ot">TRUE</span>,</a>
<a class="sourceLine" id="cb9-3" data-line-number="3">  <span class="dt">imputation.reference.panel=</span><span class="st">&quot;apps@hrc-r1.1&quot;</span>,</a>
<a class="sourceLine" id="cb9-4" data-line-number="4">  <span class="dt">imputation.phasing.method=</span><span class="st">&quot;shapeit&quot;</span>,</a>
<a class="sourceLine" id="cb9-5" data-line-number="5">  <span class="dt">imputation.population=</span><span class="st">&quot;eur&quot;</span></a>
<a class="sourceLine" id="cb9-6" data-line-number="6">)</a></code></pre></div>
<h2 id="perform-data-import">Perform data import</h2>
<p>The <code>do.import</code> function requires the paths to the respective genotyping and DNA methylation data, as well as a sample annotation sheet as discussed earlier. You’ll have to specify the paths to the corresponding <em>IDAT</em> and <em>plink</em> files. Additionally, you have to specify the sample identifier column in the sample annotation sheet that determines the samples in both the genotyping and DNA methylation data. For larger files, we recommend to activate the option to store large matrices on disk rather than in main memory (<code>hdf5dump</code>).</p>
<div class="sourceCode" id="cb10"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb10-1" data-line-number="1">idat.dir &lt;-<span class="st"> &quot;/DEEP_fhgfs/projects/mscherer/data/EPIC/CEDAR/idats/&quot;</span></a>
<a class="sourceLine" id="cb10-2" data-line-number="2">plink.dir &lt;-<span class="st"> &quot;/DEEP_fhgfs/projects/mscherer/data/450K/CEDAR/publication/139.165.108.18/srv/genmol/permanent/1be6993fe41c12a051c9244d67c91da2be49e5dd26a6cd79f442bc006971e2ef/CEDAR_GENO/&quot;</span></a>
<a class="sourceLine" id="cb10-3" data-line-number="3">anno.sheet &lt;-<span class="st"> &quot;/DEEP_fhgfs/projects/mscherer/data/EPIC/CEDAR/annotation/sample_annotation_IL_IPC_example.tsv&quot;</span></a>
<a class="sourceLine" id="cb10-4" data-line-number="4"><span class="kw">qtl.setOption</span>(<span class="dt">hdf5dump=</span><span class="ot">TRUE</span>)</a>
<a class="sourceLine" id="cb10-5" data-line-number="5"><span class="kw">setHDF5DumpDir</span>(<span class="st">&quot;/DEEP_fhgfs/projects/mscherer/deep/tmp/&quot;</span>)</a>
<a class="sourceLine" id="cb10-6" data-line-number="6">imp.data &lt;-<span class="st"> </span><span class="kw">do.import</span>(<span class="dt">data.location =</span> <span class="kw">c</span>(<span class="dt">idat.dir=</span>idat.dir,<span class="dt">geno.dir=</span>plink.dir),</a>
<a class="sourceLine" id="cb10-7" data-line-number="7">                      <span class="dt">s.anno =</span> anno.sheet,</a>
<a class="sourceLine" id="cb10-8" data-line-number="8">                      <span class="dt">s.id.col =</span> <span class="st">&quot;ind_IPC&quot;</span>,</a>
<a class="sourceLine" id="cb10-9" data-line-number="9">                      <span class="dt">tab.sep =</span> <span class="st">&quot;</span><span class="ch">\t</span><span class="st">&quot;</span>)</a></code></pre></div>
<pre><code>## 2020-05-29 13:16:34     1.3  STATUS STARTED Import methQTL data
## 2020-05-29 13:16:34     1.3  STATUS     STARTED Processing genotyping data
## 2020-05-29 13:16:42     1.7    INFO         Loading system default for option &#39;plink.path&#39;
## 2020-05-29 13:16:53     2.2  STATUS         STARTED Compute genotype PCA
## 2020-05-29 13:16:54     2.2  STATUS         COMPLETED Compute genotype PCA
## 2020-05-29 13:16:55     2.2  STATUS     COMPLETED Processing genotyping data
## 2020-05-29 13:16:55     2.2  STATUS     STARTED Processing DNA methylation data
## 2020-05-29 13:16:56     2.2  STATUS         STARTED Loading Data from IDAT Files
## 2020-05-29 13:16:56     2.2    INFO             Added column barcode to the provided sample annotation table
## 2020-05-29 13:16:57     2.2    INFO             Detected platform: MethylationEPIC
## 2020-05-29 13:17:19     2.3  STATUS         COMPLETED Loading Data from IDAT Files
## 2020-05-29 13:18:41     2.5  STATUS         STARTED Preprocessing
## 2020-05-29 13:18:41     2.5    INFO             Number of cores: 1
## 2020-05-29 13:18:41     2.5  STATUS             STARTED Filtering Procedures I
## 2020-05-29 13:18:43     2.5  STATUS                 STARTED Removal of SNP-enriched Sites
## 2020-05-29 13:18:43     2.5  STATUS                     Removed 139721 sites using SNP criterion &quot;any&quot;
## 2020-05-29 13:18:44     2.5  STATUS                     Saved removed sites to /local/tmp/RtmpH875KL/rnbeads_preprocessing/preprocessing_data/removed_sites_snp.csv
## 2020-05-29 13:18:44     2.5  STATUS                     Added a corresponding section to the report
## 2020-05-29 13:18:44     2.5  STATUS                 COMPLETED Removal of SNP-enriched Sites
## 2020-05-29 13:18:44     2.5  STATUS                 STARTED Removal of Cross-reactive Probes
## 2020-05-29 13:18:44     2.5  STATUS                     Removed 34264 sites
## 2020-05-29 13:18:44     2.5  STATUS                     Saved removed sites to /local/tmp/RtmpH875KL/rnbeads_preprocessing/preprocessing_data/removed_sites_cross_reactive.csv
## 2020-05-29 13:18:44     2.5  STATUS                     Added a corresponding section to the report
## 2020-05-29 13:18:44     2.5  STATUS                 COMPLETED Removal of Cross-reactive Probes
## 2020-05-29 13:18:44     2.5    INFO                 Working with a p-value threshold of 0.05
## 2020-05-29 13:18:46     2.6  STATUS                 STARTED Greedycut
## 2020-05-29 13:19:06     2.8  STATUS                     Calculated a total of 1055 iterations
## 2020-05-29 13:19:06     2.8    INFO                     Optimal number of iterations is 1055
## 2020-05-29 13:19:10     2.8  STATUS                     Created ROC plot
## 2020-05-29 13:19:14     2.8  STATUS                     Created line plots for matrix dimensions and other statistics
## 2020-05-29 13:19:14     2.8  STATUS                     Saved removed sites to /local/tmp/RtmpH875KL/rnbeads_preprocessing/preprocessing_data/removed_sites_greedycut.csv
## 2020-05-29 13:19:14     2.8  STATUS                 COMPLETED Greedycut
## 2020-05-29 13:19:14     2.8  STATUS                 Retained 8 samples and 691856 sites
## 2020-05-29 13:19:14     2.8  STATUS             COMPLETED Filtering Procedures I
## 2020-05-29 13:19:14     2.8  STATUS             STARTED Summary of Filtering Procedures I
## 2020-05-29 13:19:14     2.8  STATUS                 Created summary table of removed sites, samples and unreliable measurements
## 2020-05-29 13:19:16     2.8  STATUS                 Added summary table of removed and retained items
## 2020-05-29 13:19:16     2.8    INFO                 Subsampling 866895 sites for plotting density distributions
## 2020-05-29 13:19:16     2.8  STATUS                 Constructed sequences of removed and retained methylation values
## 2020-05-29 13:19:30     2.8  STATUS                 Added comparison between removed and retained beta values
## 2020-05-29 13:19:30     2.8  STATUS             COMPLETED Summary of Filtering Procedures I
## 2020-05-29 13:19:30     2.8  STATUS             STARTED Manipulating the object
## 2020-05-29 13:20:29     2.8  STATUS                 Removed 175039 sites (probes)
## 2020-05-29 13:20:29     2.8    INFO                 Retained 691856 sites and 8 samples
## 2020-05-29 13:20:29     2.8  STATUS             COMPLETED Manipulating the object
## 2020-05-29 13:20:29     2.8  STATUS             STARTED Normalization Procedure
## 2020-05-29 13:21:09     3.5  STATUS                 Performed normalization with method wm.dasen
## 2020-05-29 13:22:05     3.2  STATUS                 Performed normalization with method &quot;wm.dasen&quot;
## 2020-05-29 13:22:21     3.5  STATUS                 Added comparison between non-normalized and normalized beta values
## 2020-05-29 13:22:22     3.5  STATUS                 Added histogram of observed beta shifts (magnitude of correction)
## 2020-05-29 13:22:24     3.5  STATUS                 Added 2D histogram of observed beta values and shifts
## 2020-05-29 13:22:24     3.5  STATUS                 Added normalization section
## 2020-05-29 13:22:24     3.5  STATUS             COMPLETED Normalization Procedure
## 2020-05-29 13:22:24     3.5  STATUS             STARTED Filtering Procedures II
## 2020-05-29 13:22:26     3.5  STATUS                 STARTED Probe Context Removal
## 2020-05-29 13:22:26     3.5  STATUS                     Removed 1196 probe(s) having not acceptable context
## 2020-05-29 13:22:26     3.5  STATUS                     Saved removed sites to /local/tmp/RtmpH875KL/rnbeads_preprocessing/preprocessing_data/removed_sites_context.csv
## 2020-05-29 13:22:26     3.5  STATUS                     Added a corresponding section to the report
## 2020-05-29 13:22:26     3.5  STATUS                 COMPLETED Probe Context Removal
## 2020-05-29 13:22:26     3.5  STATUS                 STARTED Removal of Sites on Sex Chromosomes
## 2020-05-29 13:22:26     3.5  STATUS                     Removed 16598 site(s) on sex chromosomes
## 2020-05-29 13:22:26     3.5  STATUS                     Saved removed sites to /local/tmp/RtmpH875KL/rnbeads_preprocessing/preprocessing_data/removed_sites_sex.csv
## 2020-05-29 13:22:26     3.5  STATUS                     Added a corresponding section to the report
## 2020-05-29 13:22:26     3.5  STATUS                 COMPLETED Removal of Sites on Sex Chromosomes
## 2020-05-29 13:22:26     3.5  STATUS                 STARTED Missing Value Removal
## 2020-05-29 13:22:26     3.5  STATUS                     Using a sample quantile threshold of 0
## 2020-05-29 13:22:26     3.5  STATUS                     Removed 44 site(s) with too many missing values
## 2020-05-29 13:22:26     3.5  STATUS                     Saved removed sites to /local/tmp/RtmpH875KL/rnbeads_preprocessing/preprocessing_data/removed_sites_na.csv
## 2020-05-29 13:22:30     3.5  STATUS                     Added a corresponding section to the report
## 2020-05-29 13:22:30     3.5  STATUS                 COMPLETED Missing Value Removal
## 2020-05-29 13:22:30     3.5  STATUS                 Retained 8 samples and 674018 sites
## 2020-05-29 13:22:30     3.5  STATUS             COMPLETED Filtering Procedures II
## 2020-05-29 13:22:30     3.5  STATUS             STARTED Summary of Filtering Procedures II
## 2020-05-29 13:22:30     3.5  STATUS                 Created summary table of removed sites, samples and unreliable measurements
## 2020-05-29 13:22:31     3.5  STATUS                 Added summary table of removed and retained items
## 2020-05-29 13:22:32     3.5    INFO                 Subsampling 691856 sites for plotting density distributions
## 2020-05-29 13:22:32     3.5  STATUS                 Constructed sequences of removed and retained methylation values
## 2020-05-29 13:22:41     3.3  STATUS                 Added comparison between removed and retained beta values
## 2020-05-29 13:22:41     3.3  STATUS             COMPLETED Summary of Filtering Procedures II
## 2020-05-29 13:22:41     3.3  STATUS             STARTED Manipulating the object## 2020-05-29 13:23:37     3.3  STATUS                 Removed 17838 sites (probes)
## 2020-05-29 13:23:37     3.3    INFO                 Retained 674018 sites and 8 samples
## 2020-05-29 13:23:37     3.3  STATUS             COMPLETED Manipulating the object
## 2020-05-29 13:23:38     3.3    INFO             No missing values present, imputation skipped
## 2020-05-29 13:23:38     3.3  STATUS         COMPLETED Preprocessing
## 2020-05-29 13:24:07     3.3  STATUS         STARTED Removing 870 CpGs overlapping with SNPs
## 2020-05-29 13:25:05     3.3  STATUS         COMPLETED Removing 870 CpGs overlapping with SNPs
## 2020-05-29 13:25:09     3.3  STATUS     COMPLETED Processing DNA methylation data
## 2020-05-29 13:25:09     3.3  STATUS COMPLETED Import methQTL data</code></pre>
<p>For imputed data, no further processing is performed on the genotyping data and the dosage values are used as they are:</p>
<div class="sourceCode" id="cb12"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb12-1" data-line-number="1">idat.dir &lt;-<span class="st"> &quot;/DEEP_fhgfs/projects/mscherer/data/EPIC/CEDAR/idats/&quot;</span></a>
<a class="sourceLine" id="cb12-2" data-line-number="2">geno.dir &lt;-<span class="st"> &quot;/DEEP_fhgfs/projects/mscherer/data/450K/CEDAR/publication/139.165.108.18/srv/genmol/permanent/1be6993fe41c12a051c9244d67c91da2be49e5dd26a6cd79f442bc006971e2ef/CEDAR_GENO_IMPUTED/&quot;</span></a>
<a class="sourceLine" id="cb12-3" data-line-number="3">anno.sheet &lt;-<span class="st"> &quot;/DEEP_fhgfs/projects/mscherer/data/EPIC/CEDAR/annotation/sample_annotation_IL_IPC_example.tsv&quot;</span></a>
<a class="sourceLine" id="cb12-4" data-line-number="4"><span class="kw">qtl.setOption</span>(<span class="dt">hdf5dump=</span><span class="ot">TRUE</span>)</a>
<a class="sourceLine" id="cb12-5" data-line-number="5">imp.data &lt;-<span class="st"> </span><span class="kw">do.import</span>(<span class="dt">data.location =</span> <span class="kw">c</span>(<span class="dt">idat.dir=</span>idat.dir,<span class="dt">geno.dir=</span>geno.dir),</a>
<a class="sourceLine" id="cb12-6" data-line-number="6">                      <span class="dt">s.anno =</span> anno.sheet,</a>
<a class="sourceLine" id="cb12-7" data-line-number="7">                      <span class="dt">s.id.col =</span> <span class="st">&quot;ind_IPC&quot;</span>,</a>
<a class="sourceLine" id="cb12-8" data-line-number="8">                      <span class="dt">tab.sep =</span> <span class="st">&quot;</span><span class="ch">\t</span><span class="st">&quot;</span>,</a>
<a class="sourceLine" id="cb12-9" data-line-number="9">                      <span class="dt">out.folder =</span> <span class="kw">getwd</span>())</a></code></pre></div>
<p>Please note that the <code>recode.allele.frequencies</code> option specifies, if, according to the cohort analyzed, SNP reference and alternative allele are to be recoded according to the allele frequencies found. Alternatively, a path to a local version of dbSNP <span class="citation" data-cites="Sherry2001">(Sherry et al. 2001)</span> can be provided through <code>db.snp.ref</code>, and reference/alternative allele information will be automatically parsed from the database. This is especially crucial, if imputation is to be performed, since the Michigan Imputation Server is sensitive to reference mismatches. <code>recode.allele.frequencies</code> and <code>db.snp.ref</code> are mutually exclusive options.</p>
<h1 id="methqtl-calling">methQTL calling</h1>
<p>Although <em>methQTL</em> conceptually splits the methQTL calling into two steps ((i) compute correlation block, (ii) call methQTL per correlation block), only a single function call is needed. The function only requires the input <code>methQTLInput</code> object produced in the previous step, but further options, such as covariates and the p-value cutoff can be directly specified as a function parameter, or as global parameters using <code>?qtl.setOption</code>.</p>
<div class="sourceCode" id="cb13"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb13-1" data-line-number="1">meth.qtl.res &lt;-<span class="st"> </span><span class="kw">do.methQTL</span>(imp.data,)</a></code></pre></div>
<pre><code>## 2020-05-29 13:25:09     3.3    INFO Loading default option setting
## 2020-05-29 13:25:10     3.3  STATUS STARTED Imputation procedure knn 
## 2020-05-29 13:25:18     3.3  STATUS COMPLETED Imputation procedure knn 
## 
## 2020-05-29 13:25:20     3.3  STATUS STARTED Computing methQTLs
## 2020-05-29 13:25:20     3.3  STATUS     STARTED Computing methQTL for chromosome chr1
## 2020-05-29 13:25:20     3.3  STATUS         STARTED Compute correlation blocks
## 2020-05-29 13:25:20     3.3    INFO             Split workload, since facing 66034 CpGs (Maximum is 40000 )
## 2020-05-29 13:25:20     3.3  STATUS             STARTED Compute correlation blocks
## 2020-05-29 13:25:20     3.3  STATUS                 STARTED Compute correlation matrix
## 2020-05-29 13:25:54    19.6  STATUS                 COMPLETED Compute correlation matrix
## 2020-05-29 13:36:45     3.8  STATUS                 STARTED Compute pairwise distances
## 2020-05-29 13:37:11    13.0  STATUS                 COMPLETED Compute pairwise distances
## 2020-05-29 13:38:00    10.2  STATUS                 STARTED Weight distances
## 2020-05-29 13:43:57    27.2  STATUS                 COMPLETED Weight distances
## 2020-05-29 13:43:58    15.8  STATUS                 STARTED Compute graph
## 2020-05-29 13:44:51    32.1  STATUS                 COMPLETED Compute graph
## 2020-05-29 13:44:51    32.1  STATUS                 STARTED Compute clustering
## 2020-05-29 13:44:53    32.1  STATUS                 COMPLETED Compute clustering
## 2020-05-29 13:44:53    32.1  STATUS             COMPLETED Compute correlation blocks
## 2020-05-29 13:44:53    32.1  STATUS             STARTED Compute correlation blocks
## 2020-05-29 13:44:53    32.1  STATUS                 STARTED Compute correlation matrix
## 2020-05-29 13:45:25    48.3  STATUS                 COMPLETED Compute correlation matrix
## 2020-05-29 13:56:15    38.2  STATUS                 STARTED Compute pairwise distances
## 2020-05-29 13:56:40    22.4  STATUS                 COMPLETED Compute pairwise distances
## 2020-05-29 13:57:27    10.2  STATUS                 STARTED Weight distances
## 2020-05-29 14:03:44    37.4  STATUS                 COMPLETED Weight distances
## 2020-05-29 14:03:46    19.1  STATUS                 STARTED Compute graph
## 2020-05-29 14:04:35    43.4  STATUS                 COMPLETED Compute graph
## 2020-05-29 14:04:35    43.4  STATUS                 STARTED Compute clustering
## 2020-05-29 14:04:36    43.4  STATUS                 COMPLETED Compute clustering
## 2020-05-29 14:04:36    43.4  STATUS             COMPLETED Compute correlation blocks
## 2020-05-29 14:04:36    43.4  STATUS             STARTED Compute methQTL per correlation block
## 2020-05-29 14:04:36    43.4  STATUS                 STARTED Setting up Multicore
## 2020-05-29 14:04:36    43.4    INFO                     Using 1 cores
## 2020-05-29 14:04:36    43.4  STATUS                 COMPLETED Setting up Multicore
## 2020-05-29 14:04:37    43.4    INFO                 No SNP closer than 500000
## 2020-05-29 14:54:47     6.1    INFO                 No SNP closer than 500000
## 2020-05-29 14:54:47     6.1    INFO                 No SNP closer than 500000
## 2020-05-29 14:54:47     6.1    INFO                 No SNP closer than 500000
## 2020-05-29 14:54:48     6.1    INFO                 No SNP closer than 500000
## 2020-05-29 14:54:48     6.1    INFO                 No SNP closer than 500000
## 2020-05-29 14:54:48     6.1    INFO                 No SNP closer than 500000
## 2020-05-29 14:54:49     6.1    INFO                 No SNP closer than 500000
## 2020-05-29 14:54:49     6.1    INFO                 No SNP closer than 500000
## 2020-05-29 14:54:49     6.1    INFO                 No SNP closer than 500000
## 2020-05-29 14:54:50     6.1    INFO                 No SNP closer than 500000
## 2020-05-29 14:54:50     6.1    INFO                 No SNP closer than 500000
## 2020-05-29 15:40:36     6.1  STATUS             COMPLETED Compute methQTL per correlation block
## 2020-05-29 15:40:36     6.1  STATUS         COMPLETED Compute correlation blocks
## 2020-05-29 15:40:36     6.1  STATUS         STARTED Computing methQTL for chromosome chr2
## 2020-05-29 15:40:36     6.1  STATUS             STARTED Compute correlation blocks
## 2020-05-29 15:40:36     6.1    INFO                 Split workload, since facing 52088 CpGs (Maximum is 40000 )
## 2020-05-29 15:40:36     6.1  STATUS                 STARTED Compute correlation blocks
## 2020-05-29 15:40:36     6.1  STATUS                     STARTED Compute correlation matrix
## 2020-05-29 15:40:55    16.3  STATUS                     COMPLETED Compute correlation matrix
## 2020-05-29 15:47:45    11.2  STATUS                     STARTED Compute pairwise distances
## 2020-05-29 15:48:01     9.2  STATUS                     COMPLETED Compute pairwise distances
## 2020-05-29 15:48:38     7.1  STATUS                     STARTED Weight distances
## 2020-05-29 15:52:24    16.8  STATUS                     COMPLETED Weight distances
## 2020-05-29 15:52:25    12.1  STATUS                     STARTED Compute graph
## 2020-05-29 15:52:55    34.3  STATUS                     COMPLETED Compute graph
## 2020-05-29 15:52:55    34.3  STATUS                     STARTED Compute clustering
## 2020-05-29 15:52:56    34.3  STATUS                     COMPLETED Compute clustering
## 2020-05-29 15:52:56    34.3  STATUS                 COMPLETED Compute correlation blocks
## 2020-05-29 15:52:56    34.3  STATUS                 STARTED Compute correlation blocks
## 2020-05-29 15:52:56    34.3  STATUS                     STARTED Compute correlation matrix
## 2020-05-29 15:53:15    39.4  STATUS                     COMPLETED Compute correlation matrix
## 2020-05-29 16:00:02    18.3  STATUS                     STARTED Compute pairwise distances
## 2020-05-29 16:00:18    14.7  STATUS                     COMPLETED Compute pairwise distances
## 2020-05-29 16:00:47     7.1  STATUS                     STARTED Weight distances
## 2020-05-29 16:04:28    17.4  STATUS                     COMPLETED Weight distances
## 2020-05-29 16:04:29    12.1  STATUS                     STARTED Compute graph
## 2020-05-29 16:04:59    34.3  STATUS                     COMPLETED Compute graph
## 2020-05-29 16:04:59    34.3  STATUS                     STARTED Compute clustering
## 2020-05-29 16:05:00    34.3  STATUS                     COMPLETED Compute clustering
## 2020-05-29 16:05:00    34.3  STATUS                 COMPLETED Compute correlation blocks
## 2020-05-29 16:05:01    34.3  STATUS                 STARTED Compute methQTL per correlation block
## 2020-05-29 16:05:01    34.3  STATUS                     STARTED Setting up Multicore
## 2020-05-29 16:05:01    34.3    INFO                         Using 1 cores
## 2020-05-29 16:05:01    34.3  STATUS                     COMPLETED Setting up Multicore
## 2020-05-29 17:39:39     6.1  STATUS                 COMPLETED Compute methQTL per correlation block
## 2020-05-29 17:39:39     6.1  STATUS             COMPLETED Compute correlation blocks
## 2020-05-29 17:39:39     6.1  STATUS             STARTED Computing methQTL for chromosome chr3
## 2020-05-29 17:39:39     6.1  STATUS                 STARTED Compute correlation blocks
## 2020-05-29 17:39:39     6.1  STATUS                     STARTED Compute correlation matrix
## 2020-05-29 17:40:26    29.8  STATUS                     COMPLETED Compute correlation matrix
## 2020-05-29 17:56:06    18.6  STATUS                     STARTED Compute pairwise distances
## 2020-05-29 17:56:37    18.6  STATUS                     COMPLETED Compute pairwise distances
## 2020-05-29 17:57:44    13.9  STATUS                     STARTED Weight distances
## 2020-05-29 18:06:16    39.0  STATUS                     COMPLETED Weight distances
## 2020-05-29 18:06:17    25.7  STATUS                     STARTED Compute graph
## 2020-05-29 18:10:43    49.4  STATUS                     COMPLETED Compute graph
## 2020-05-29 18:10:43    49.4  STATUS                     STARTED Compute clustering
## 2020-05-29 18:10:44    49.4  STATUS                     COMPLETED Compute clustering
## 2020-05-29 18:10:44    49.4  STATUS                 COMPLETED Compute correlation blocks
## 2020-05-29 18:10:45    49.4  STATUS                 STARTED Compute methQTL per correlation block
## 2020-05-29 18:10:45    49.4  STATUS                     STARTED Setting up Multicore
## 2020-05-29 18:10:45    49.4    INFO                         Using 1 cores
## 2020-05-29 18:10:45    49.4  STATUS                     COMPLETED Setting up Multicore
## 2020-05-29 19:24:07     6.1  STATUS                 COMPLETED Compute methQTL per correlation block
## 2020-05-29 19:24:07     6.1  STATUS             COMPLETED Computing methQTL for chromosome chr3
## 2020-05-29 19:24:07     6.1  STATUS             STARTED Computing methQTL for chromosome chr4
## 2020-05-29 19:24:07     6.1  STATUS                 STARTED Compute correlation blocks
## 2020-05-29 19:24:07     6.1  STATUS                     STARTED Compute correlation matrix
## 2020-05-29 19:24:32    18.6  STATUS                     COMPLETED Compute correlation matrix
## 2020-05-29 19:32:52    25.0  STATUS                     STARTED Compute pairwise distances
## 2020-05-29 19:33:10    10.9  STATUS                     COMPLETED Compute pairwise distances
## 2020-05-29 19:33:48     8.3  STATUS                     STARTED Weight distances
## 2020-05-29 19:38:22    25.2  STATUS                     COMPLETED Weight distances
## 2020-05-29 19:38:23    14.5  STATUS                     STARTED Compute graph
## 2020-05-29 19:38:58    43.6  STATUS                     COMPLETED Compute graph
## 2020-05-29 19:38:58    43.6  STATUS                     STARTED Compute clustering
## 2020-05-29 19:38:59    43.6  STATUS                     COMPLETED Compute clustering
## 2020-05-29 19:38:59    43.6  STATUS                 COMPLETED Compute correlation blocks
## 2020-05-29 19:38:59    43.6  STATUS                 STARTED Compute methQTL per correlation block
## 2020-05-29 19:38:59    43.6  STATUS                     STARTED Setting up Multicore
## 2020-05-29 19:38:59    43.6    INFO                         Using 1 cores
## 2020-05-29 19:38:59    43.6  STATUS                     COMPLETED Setting up Multicore
## 2020-05-29 20:38:20     6.1  STATUS                 COMPLETED Compute methQTL per correlation block
## 2020-05-29 20:38:20     6.1  STATUS             COMPLETED Computing methQTL for chromosome chr4
## 2020-05-29 20:38:20     6.1  STATUS             STARTED Computing methQTL for chromosome chr5
## 2020-05-29 20:38:20     6.1  STATUS                 STARTED Compute correlation blocks
## 2020-05-29 20:38:20     6.1  STATUS                     STARTED Compute correlation matrix
## 2020-05-29 20:39:00    25.3  STATUS                     COMPLETED Compute correlation matrix
## 2020-05-29 20:51:48    16.5  STATUS                     STARTED Compute pairwise distances
## 2020-05-29 20:52:14    15.5  STATUS                     COMPLETED Compute pairwise distances
## 2020-05-29 20:53:10    11.6  STATUS                     STARTED Weight distances
## 2020-05-29 21:00:04    30.5  STATUS                     COMPLETED Weight distances
## 2020-05-29 21:00:06    21.2  STATUS                     STARTED Compute graph
## 2020-05-29 21:01:06    40.4  STATUS                     COMPLETED Compute graph
## 2020-05-29 21:01:06    40.4  STATUS                     STARTED Compute clustering
## 2020-05-29 21:01:07    40.4  STATUS                     COMPLETED Compute clustering
## 2020-05-29 21:01:07    40.4  STATUS                 COMPLETED Compute correlation blocks
## 2020-05-29 21:01:08    40.4  STATUS                 STARTED Compute methQTL per correlation block
## 2020-05-29 21:01:08    40.4  STATUS                     STARTED Setting up Multicore
## 2020-05-29 21:01:08    40.4    INFO                         Using 1 cores
## 2020-05-29 21:01:08    40.4  STATUS                     COMPLETED Setting up Multicore
## 2020-05-29 21:24:46     6.1    INFO                     No SNP closer than 500000
## 2020-05-29 21:24:46     6.1    INFO                     No SNP closer than 500000
## 2020-05-29 22:05:42     6.1  STATUS                 COMPLETED Compute methQTL per correlation block
## 2020-05-29 22:05:42     6.1  STATUS             COMPLETED Computing methQTL for chromosome chr5
## 2020-05-29 22:05:42     6.1  STATUS             STARTED Computing methQTL for chromosome chr6
## 2020-05-29 22:05:42     6.1  STATUS                 STARTED Compute correlation blocks
## 2020-05-29 22:05:42     6.1    INFO                     Split workload, since facing 41991 CpGs (Maximum is 40000 )
## 2020-05-29 22:05:42     6.1  STATUS                     STARTED Compute correlation blocks
## 2020-05-29 22:05:42     6.1  STATUS                         STARTED Compute correlation matrix
## 2020-05-29 22:05:54    12.7  STATUS                         COMPLETED Compute correlation matrix
## 2020-05-29 22:09:57     7.3  STATUS                         STARTED Compute pairwise distances
## 2020-05-29 22:10:08    10.3  STATUS                         COMPLETED Compute pairwise distances
## 2020-05-29 22:10:26     6.1  STATUS                         STARTED Weight distances
## 2020-05-29 22:12:52     9.2  STATUS                         COMPLETED Weight distances
## 2020-05-29 22:12:54     9.2  STATUS                         STARTED Compute graph
## 2020-05-29 22:13:14    23.9  STATUS                         COMPLETED Compute graph
## 2020-05-29 22:13:14    23.9  STATUS                         STARTED Compute clustering
## 2020-05-29 22:13:16    23.9  STATUS                         COMPLETED Compute clustering
## 2020-05-29 22:13:16    23.9  STATUS                     COMPLETED Compute correlation blocks
## 2020-05-29 22:13:16    23.9  STATUS                     STARTED Compute correlation blocks
## 2020-05-29 22:13:16    23.9  STATUS                         STARTED Compute correlation matrix
## 2020-05-29 22:13:29    27.2  STATUS                         COMPLETED Compute correlation matrix
## 2020-05-29 22:17:41    13.0  STATUS                         STARTED Compute pairwise distances
## 2020-05-29 22:17:52    10.3  STATUS                         COMPLETED Compute pairwise distances
## 2020-05-29 22:18:11     6.1  STATUS                         STARTED Weight distances
## 2020-05-29 22:20:32    13.1  STATUS                         COMPLETED Weight distances
## 2020-05-29 22:20:34     9.2  STATUS                         STARTED Compute graph
## 2020-05-29 22:20:55    23.9  STATUS                         COMPLETED Compute graph
## 2020-05-29 22:20:55    23.9  STATUS                         STARTED Compute clustering
## 2020-05-29 22:20:56    23.9  STATUS                         COMPLETED Compute clustering
## 2020-05-29 22:20:56    23.9  STATUS                     COMPLETED Compute correlation blocks
## 2020-05-29 22:20:56    23.9  STATUS                     STARTED Compute methQTL per correlation block
## 2020-05-29 22:20:56    23.9  STATUS                         STARTED Setting up Multicore
## 2020-05-29 22:20:56    23.9    INFO                             Using 1 cores
## 2020-05-29 22:20:56    23.9  STATUS                         COMPLETED Setting up Multicore
## 2020-05-29 23:33:08     6.1  STATUS                     COMPLETED Compute methQTL per correlation block
## 2020-05-29 23:33:08     6.1  STATUS                 COMPLETED Compute correlation blocks
## 2020-05-29 23:33:08     6.1  STATUS                 STARTED Computing methQTL for chromosome chr7
## 2020-05-29 23:33:08     6.1  STATUS                     STARTED Compute correlation blocks
## 2020-05-29 23:33:08     6.1  STATUS                         STARTED Compute correlation matrix
## 2020-05-29 23:33:48    26.2  STATUS                         COMPLETED Compute correlation matrix
## 2020-05-29 23:47:00    16.3  STATUS                         STARTED Compute pairwise distances
## 2020-05-29 23:47:27    16.1  STATUS                         COMPLETED Compute pairwise distances
## 2020-05-29 23:48:25    12.1  STATUS                         STARTED Weight distances
## 2020-05-29 23:55:39    28.6  STATUS                         COMPLETED Weight distances
## 2020-05-29 23:55:41    22.1  STATUS                         STARTED Compute graph
## 2020-05-29 23:59:15    42.2  STATUS                         COMPLETED Compute graph
## 2020-05-29 23:59:15    42.2  STATUS                         STARTED Compute clustering
## 2020-05-29 23:59:17    42.2  STATUS                         COMPLETED Compute clustering
## 2020-05-29 23:59:17    42.2  STATUS                     COMPLETED Compute correlation blocks
## 2020-05-29 23:59:17    42.2  STATUS                     STARTED Compute methQTL per correlation block
## 2020-05-29 23:59:17    42.2  STATUS                         STARTED Setting up Multicore
## 2020-05-29 23:59:17    42.2    INFO                             Using 1 cores
## 2020-05-29 23:59:17    42.2  STATUS                         COMPLETED Setting up Multicore
## 2020-05-30 00:57:27     6.1  STATUS                     COMPLETED Compute methQTL per correlation block
## 2020-05-30 00:57:27     6.1  STATUS                 COMPLETED Computing methQTL for chromosome chr7
## 2020-05-30 00:57:27     6.1  STATUS                 STARTED Computing methQTL for chromosome chr8
## 2020-05-30 00:57:27     6.1  STATUS                     STARTED Compute correlation blocks
## 2020-05-30 00:57:27     6.1  STATUS                         STARTED Compute correlation matrix
## 2020-05-30 00:57:53    19.9  STATUS                         COMPLETED Compute correlation matrix
## 2020-05-30 01:07:24    13.3  STATUS                         STARTED Compute pairwise distances
## 2020-05-30 01:07:43    13.9  STATUS                         COMPLETED Compute pairwise distances
## 2020-05-30 01:08:22     8.9  STATUS                         STARTED Weight distances
## 2020-05-30 01:13:31    27.1  STATUS                         COMPLETED Weight distances
## 2020-05-30 01:13:33    15.7  STATUS                         STARTED Compute graph
## 2020-05-30 01:14:18    29.5  STATUS                         COMPLETED Compute graph
## 2020-05-30 01:14:18    29.5  STATUS                         STARTED Compute clustering
## 2020-05-30 01:14:19    29.5  STATUS                         COMPLETED Compute clustering
## 2020-05-30 01:14:19    29.5  STATUS                     COMPLETED Compute correlation blocks
## 2020-05-30 01:14:20    29.5  STATUS                     STARTED Compute methQTL per correlation block
## 2020-05-30 01:14:20    29.5  STATUS                         STARTED Setting up Multicore
## 2020-05-30 01:14:20    29.5    INFO                             Using 1 cores
## 2020-05-30 01:14:20    29.5  STATUS                         COMPLETED Setting up Multicore
## 2020-05-30 02:09:43     6.1  STATUS                     COMPLETED Compute methQTL per correlation block
## 2020-05-30 02:09:43     6.1  STATUS                 COMPLETED Computing methQTL for chromosome chr8
## 2020-05-30 02:09:43     6.1  STATUS                 STARTED Computing methQTL for chromosome chr9
## 2020-05-30 02:09:43     6.1  STATUS                     STARTED Compute correlation blocks
## 2020-05-30 02:09:43     6.1  STATUS                         STARTED Compute correlation matrix
## 2020-05-30 02:09:55    12.7  STATUS                         COMPLETED Compute correlation matrix
## 2020-05-30 02:14:06    11.8  STATUS                         STARTED Compute pairwise distances
## 2020-05-30 02:14:17    10.0  STATUS                         COMPLETED Compute pairwise distances
## 2020-05-30 02:14:35     6.1  STATUS                         STARTED Weight distances
## 2020-05-30 02:17:02     9.2  STATUS                         COMPLETED Weight distances
## 2020-05-30 02:17:03     9.2  STATUS                         STARTED Compute graph
## 2020-05-30 02:17:23    23.8  STATUS                         COMPLETED Compute graph
## 2020-05-30 02:17:23    23.8  STATUS                         STARTED Compute clustering
## 2020-05-30 02:17:24    23.8  STATUS                         COMPLETED Compute clustering
## 2020-05-30 02:17:24    23.8  STATUS                     COMPLETED Compute correlation blocks
## 2020-05-30 02:17:24    23.8  STATUS                     STARTED Compute methQTL per correlation block
## 2020-05-30 02:17:24    23.8  STATUS                         STARTED Setting up Multicore
## 2020-05-30 02:17:24    23.8    INFO                             Using 1 cores
## 2020-05-30 02:17:24    23.8  STATUS                         COMPLETED Setting up Multicore
## 2020-05-30 02:30:53     6.1    INFO                         No SNP closer than 500000
## 2020-05-30 02:30:54     6.1    INFO                         No SNP closer than 500000
## 2020-05-30 02:30:54     6.1    INFO                         No SNP closer than 500000
## 2020-05-30 02:30:54     6.1    INFO                         No SNP closer than 500000
## 2020-05-30 02:30:55     6.1    INFO                         No SNP closer than 500000
## 2020-05-30 02:30:55     6.1    INFO                         No SNP closer than 500000
## 2020-05-30 02:30:56     6.1    INFO                         No SNP closer than 500000
## 2020-05-30 02:30:57     6.1    INFO                         No SNP closer than 500000
## 2020-05-30 02:30:58     6.1    INFO                         No SNP closer than 500000
## 2020-05-30 02:30:58     6.1    INFO                         No SNP closer than 500000
## 2020-05-30 02:30:58     6.1    INFO                         No SNP closer than 500000
## 2020-05-30 02:30:59     6.1    INFO                         No SNP closer than 500000
## 2020-05-30 02:30:59     6.1    INFO                         No SNP closer than 500000
## 2020-05-30 02:30:59     6.1    INFO                         No SNP closer than 500000
## 2020-05-30 02:31:00     6.1    INFO                         No SNP closer than 500000
## 2020-05-30 02:31:00     6.1    INFO                         No SNP closer than 500000
## 2020-05-30 02:31:05     6.1    INFO                         No SNP closer than 500000
## 2020-05-30 02:31:06     6.1    INFO                         No SNP closer than 500000
## 2020-05-30 02:31:06     6.1    INFO                         No SNP closer than 500000
## 2020-05-30 03:02:44     6.1  STATUS                     COMPLETED Compute methQTL per correlation block
## 2020-05-30 03:02:44     6.1  STATUS                 COMPLETED Computing methQTL for chromosome chr9
## 2020-05-30 03:02:44     6.1  STATUS                 STARTED Computing methQTL for chromosome chr10
## 2020-05-30 03:02:44     6.1  STATUS                     STARTED Compute correlation blocks
## 2020-05-30 03:02:44     6.1  STATUS                         STARTED Compute correlation matrix
## 2020-05-30 03:03:15    22.8  STATUS                         COMPLETED Compute correlation matrix
## 2020-05-30 03:14:35    15.0  STATUS                         STARTED Compute pairwise distances
## 2020-05-30 03:14:59    13.8  STATUS                         COMPLETED Compute pairwise distances
## 2020-05-30 03:15:56    10.4  STATUS                         STARTED Weight distances
## 2020-05-30 03:22:05    23.8  STATUS                         COMPLETED Weight distances
## 2020-05-30 03:22:06    17.9  STATUS                         STARTED Compute graph
## 2020-05-30 03:23:21    34.6  STATUS                         COMPLETED Compute graph
## 2020-05-30 03:23:21    34.6  STATUS                         STARTED Compute clustering
## 2020-05-30 03:23:22    34.6  STATUS                         COMPLETED Compute clustering
## 2020-05-30 03:23:22    34.6  STATUS                     COMPLETED Compute correlation blocks
## 2020-05-30 03:23:22    34.6  STATUS                     STARTED Compute methQTL per correlation block
## 2020-05-30 03:23:22    34.6  STATUS                         STARTED Setting up Multicore
## 2020-05-30 03:23:22    34.6    INFO                             Using 1 cores
## 2020-05-30 03:23:22    34.6  STATUS                         COMPLETED Setting up Multicore
## 2020-05-30 04:20:44     6.1  STATUS                     COMPLETED Compute methQTL per correlation block
## 2020-05-30 04:20:44     6.1  STATUS                 COMPLETED Computing methQTL for chromosome chr10
## 2020-05-30 04:20:44     6.1  STATUS                 STARTED Computing methQTL for chromosome chr11
## 2020-05-30 04:20:44     6.1  STATUS                     STARTED Compute correlation blocks
## 2020-05-30 04:20:44     6.1  STATUS                         STARTED Compute correlation matrix
## 2020-05-30 04:21:27    29.5  STATUS                         COMPLETED Compute correlation matrix
## 2020-05-30 04:36:52    18.1  STATUS                         STARTED Compute pairwise distances
## 2020-05-30 04:37:23    19.2  STATUS                         COMPLETED Compute pairwise distances
## 2020-05-30 04:38:29    13.8  STATUS                         STARTED Weight distances
## 2020-05-30 04:46:51    37.0  STATUS                         COMPLETED Weight distances
## 2020-05-30 04:46:53    25.5  STATUS                         STARTED Compute graph
## 2020-05-30 04:54:19    48.9  STATUS                         COMPLETED Compute graph
## 2020-05-30 04:54:19    48.9  STATUS                         STARTED Compute clustering
## 2020-05-30 04:54:22    48.9  STATUS                         COMPLETED Compute clustering
## 2020-05-30 04:54:22    48.9  STATUS                     COMPLETED Compute correlation blocks
## 2020-05-30 04:54:22    48.9  STATUS                     STARTED Compute methQTL per correlation block
## 2020-05-30 04:54:22    48.9  STATUS                         STARTED Setting up Multicore
## 2020-05-30 04:54:22    48.9    INFO                             Using 1 cores
## 2020-05-30 04:54:22    48.9  STATUS                         COMPLETED Setting up Multicore
## 2020-05-30 05:48:46     6.1  STATUS                     COMPLETED Compute methQTL per correlation block
## 2020-05-30 05:48:46     6.1  STATUS                 COMPLETED Computing methQTL for chromosome chr11
## 2020-05-30 05:48:46     6.1  STATUS                 STARTED Computing methQTL for chromosome chr12
## 2020-05-30 05:48:46     6.1  STATUS                     STARTED Compute correlation blocks
## 2020-05-30 05:48:46     6.1  STATUS                         STARTED Compute correlation matrix
## 2020-05-30 05:49:21    25.6  STATUS                         COMPLETED Compute correlation matrix
## 2020-05-30 06:02:33    12.3  STATUS                         STARTED Compute pairwise distances
## 2020-05-30 06:02:58    15.9  STATUS                         COMPLETED Compute pairwise distances
## 2020-05-30 06:03:56    11.8  STATUS                         STARTED Weight distances
## 2020-05-30 06:11:04    31.6  STATUS                         COMPLETED Weight distances
## 2020-05-30 06:11:05    21.6  STATUS                         STARTED Compute graph
## 2020-05-30 06:12:07    41.1  STATUS                         COMPLETED Compute graph
## 2020-05-30 06:12:07    41.1  STATUS                         STARTED Compute clustering
## 2020-05-30 06:12:08    41.1  STATUS                         COMPLETED Compute clustering
## 2020-05-30 06:12:08    41.1  STATUS                     COMPLETED Compute correlation blocks
## 2020-05-30 06:12:09    41.1  STATUS                     STARTED Compute methQTL per correlation block
## 2020-05-30 06:12:09    41.1  STATUS                         STARTED Setting up Multicore
## 2020-05-30 06:12:09    41.1    INFO                             Using 1 cores
## 2020-05-30 06:12:09    41.1  STATUS                         COMPLETED Setting up Multicore
## 2020-05-30 07:07:30     6.1  STATUS                     COMPLETED Compute methQTL per correlation block
## 2020-05-30 07:07:30     6.1  STATUS                 COMPLETED Computing methQTL for chromosome chr12
## 2020-05-30 07:07:30     6.1  STATUS                 STARTED Computing methQTL for chromosome chr13
## 2020-05-30 07:07:30     6.1  STATUS                     STARTED Compute correlation blocks
## 2020-05-30 07:07:30     6.1  STATUS                         STARTED Compute correlation matrix
## 2020-05-30 07:07:38    10.2  STATUS                         COMPLETED Compute correlation matrix
## 2020-05-30 07:10:11    16.5  STATUS                         STARTED Compute pairwise distances
## 2020-05-30 07:10:18     7.1  STATUS                         COMPLETED Compute pairwise distances
## 2020-05-30 07:10:32     6.1  STATUS                         STARTED Weight distances
## 2020-05-30 07:11:59    10.9  STATUS                         COMPLETED Weight distances
## 2020-05-30 07:12:00     8.1  STATUS                         STARTED Compute graph
## 2020-05-30 07:12:12    14.2  STATUS                         COMPLETED Compute graph
## 2020-05-30 07:12:12    14.2  STATUS                         STARTED Compute clustering
## 2020-05-30 07:12:13    14.2  STATUS                         COMPLETED Compute clustering
## 2020-05-30 07:12:13    14.2  STATUS                     COMPLETED Compute correlation blocks
## 2020-05-30 07:12:13    14.2  STATUS                     STARTED Compute methQTL per correlation block
## 2020-05-30 07:12:13    14.2  STATUS                         STARTED Setting up Multicore
## 2020-05-30 07:12:13    14.2    INFO                             Using 1 cores
## 2020-05-30 07:12:13    14.2  STATUS                         COMPLETED Setting up Multicore
## 2020-05-30 07:48:48     6.1  STATUS                     COMPLETED Compute methQTL per correlation block
## 2020-05-30 07:48:48     6.1  STATUS                 COMPLETED Computing methQTL for chromosome chr13
## 2020-05-30 07:48:48     6.1  STATUS                 STARTED Computing methQTL for chromosome chr14
## 2020-05-30 07:48:48     6.1  STATUS                     STARTED Compute correlation blocks
## 2020-05-30 07:48:48     6.1  STATUS                         STARTED Compute correlation matrix
## 2020-05-30 07:49:05    14.4  STATUS                         COMPLETED Compute correlation matrix
## 2020-05-30 07:54:20     7.4  STATUS                         STARTED Compute pairwise distances
## 2020-05-30 07:54:35     7.8  STATUS                         COMPLETED Compute pairwise distances
## 2020-05-30 07:55:04     6.2  STATUS                         STARTED Weight distances
## 2020-05-30 07:58:10    14.5  STATUS                         COMPLETED Weight distances
## 2020-05-30 07:58:11    10.4  STATUS                         STARTED Compute graph
## 2020-05-30 07:58:36    29.4  STATUS                         COMPLETED Compute graph
## 2020-05-30 07:58:36    29.4  STATUS                         STARTED Compute clustering
## 2020-05-30 07:58:37    29.4  STATUS                         COMPLETED Compute clustering
## 2020-05-30 07:58:37    29.4  STATUS                     COMPLETED Compute correlation blocks
## 2020-05-30 07:58:37    29.4  STATUS                     STARTED Compute methQTL per correlation block
## 2020-05-30 07:58:37    29.4  STATUS                         STARTED Setting up Multicore
## 2020-05-30 07:58:37    29.4    INFO                             Using 1 cores
## 2020-05-30 07:58:37    29.4  STATUS                         COMPLETED Setting up Multicore
## 2020-05-30 08:34:09     6.1    INFO                         No SNP closer than 500000
## 2020-05-30 08:34:28     6.1  STATUS                     COMPLETED Compute methQTL per correlation block
## 2020-05-30 08:34:28     6.1  STATUS                 COMPLETED Computing methQTL for chromosome chr14
## 2020-05-30 08:34:28     6.1  STATUS                 STARTED Computing methQTL for chromosome chr15
## 2020-05-30 08:34:28     6.1  STATUS                     STARTED Compute correlation blocks
## 2020-05-30 08:34:28     6.1  STATUS                         STARTED Compute correlation matrix
## 2020-05-30 08:34:41    13.8  STATUS                         COMPLETED Compute correlation matrix
## 2020-05-30 08:39:53     6.4  STATUS                         STARTED Compute pairwise distances
## 2020-05-30 08:40:04     8.9  STATUS                         COMPLETED Compute pairwise distances
## 2020-05-30 08:40:28     6.1  STATUS                         STARTED Weight distances
## 2020-05-30 08:43:21    12.8  STATUS                         COMPLETED Weight distances
## 2020-05-30 08:43:23     9.9  STATUS                         STARTED Compute graph
## 2020-05-30 08:43:45    27.3  STATUS                         COMPLETED Compute graph
## 2020-05-30 08:43:45    27.3  STATUS                         STARTED Compute clustering
## 2020-05-30 08:43:46    27.3  STATUS                         COMPLETED Compute clustering
## 2020-05-30 08:43:46    27.3  STATUS                     COMPLETED Compute correlation blocks
## 2020-05-30 08:43:46    27.3  STATUS                     STARTED Compute methQTL per correlation block
## 2020-05-30 08:43:46    27.3  STATUS                         STARTED Setting up Multicore
## 2020-05-30 08:43:46    27.3    INFO                             Using 1 cores
## 2020-05-30 08:43:46    27.3  STATUS                         COMPLETED Setting up Multicore
## 2020-05-30 08:43:49    27.3    INFO                         No SNP closer than 500000
## 2020-05-30 08:43:49    27.3    INFO                         No SNP closer than 500000
## 2020-05-30 08:43:50    27.3    INFO                         No SNP closer than 500000
## 2020-05-30 08:43:50    27.3    INFO                         No SNP closer than 500000
## 2020-05-30 08:43:50    27.3    INFO                         No SNP closer than 500000
## 2020-05-30 08:43:51    27.3    INFO                         No SNP closer than 500000
## 2020-05-30 08:43:51    27.3    INFO                         No SNP closer than 500000
## 2020-05-30 08:43:51    27.3    INFO                         No SNP closer than 500000
## 2020-05-30 08:43:52    27.3    INFO                         No SNP closer than 500000
## 2020-05-30 09:19:46     6.1  STATUS                     COMPLETED Compute methQTL per correlation block
## 2020-05-30 09:19:46     6.1  STATUS                 COMPLETED Computing methQTL for chromosome chr15
## 2020-05-30 09:19:46     6.1  STATUS                 STARTED Computing methQTL for chromosome chr16
## 2020-05-30 09:19:46     6.1  STATUS                     STARTED Compute correlation blocks
## 2020-05-30 09:19:46     6.1  STATUS                         STARTED Compute correlation matrix
## 2020-05-30 09:20:10    19.0  STATUS                         COMPLETED Compute correlation matrix
## 2020-05-30 09:28:40    12.6  STATUS                         STARTED Compute pairwise distances
## 2020-05-30 09:28:59    11.2  STATUS                         COMPLETED Compute pairwise distances
## 2020-05-30 09:29:39     8.5  STATUS                         STARTED Weight distances
## 2020-05-30 09:34:22    20.2  STATUS                         COMPLETED Weight distances
## 2020-05-30 09:34:24    15.0  STATUS                         STARTED Compute graph
## 2020-05-30 09:35:02    45.1  STATUS                         COMPLETED Compute graph
## 2020-05-30 09:35:02    45.1  STATUS                         STARTED Compute clustering
## 2020-05-30 09:35:04    45.1  STATUS                         COMPLETED Compute clustering
## 2020-05-30 09:35:04    45.1  STATUS                     COMPLETED Compute correlation blocks
## 2020-05-30 09:35:04    45.1  STATUS                     STARTED Compute methQTL per correlation block
## 2020-05-30 09:35:04    45.1  STATUS                         STARTED Setting up Multicore
## 2020-05-30 09:35:04    45.1    INFO                             Using 1 cores
## 2020-05-30 09:35:04    45.1  STATUS                         COMPLETED Setting up Multicore
## 2020-05-30 09:48:04     6.1    INFO                         No SNP closer than 500000
## 2020-05-30 09:48:04     6.1    INFO                         No SNP closer than 500000
## 2020-05-30 09:48:05     6.1    INFO                         No SNP closer than 500000
## 2020-05-30 09:48:05     6.1    INFO                         No SNP closer than 500000
## 2020-05-30 09:48:05     6.1    INFO                         No SNP closer than 500000
## 2020-05-30 09:48:06     6.1    INFO                         No SNP closer than 500000
## 2020-05-30 09:48:06     6.1    INFO                         No SNP closer than 500000
## 2020-05-30 09:48:06     6.1    INFO                         No SNP closer than 500000
## 2020-05-30 10:07:18     6.1  STATUS                     COMPLETED Compute methQTL per correlation block
## 2020-05-30 10:07:18     6.1  STATUS                 COMPLETED Computing methQTL for chromosome chr16
## 2020-05-30 10:07:18     6.1  STATUS                 STARTED Computing methQTL for chromosome chr17
## 2020-05-30 10:07:18     6.1  STATUS                     STARTED Compute correlation blocks
## 2020-05-30 10:07:18     6.1  STATUS                         STARTED Compute correlation matrix
## 2020-05-30 10:07:52    24.9  STATUS                         COMPLETED Compute correlation matrix
## 2020-05-30 10:20:19    34.7  STATUS                         STARTED Compute pairwise distances
## 2020-05-30 10:20:45    15.6  STATUS                         COMPLETED Compute pairwise distances
## 2020-05-30 10:21:39    11.5  STATUS                         STARTED Weight distances
## 2020-05-30 10:28:29    27.8  STATUS                         COMPLETED Weight distances
## 2020-05-30 10:28:31    20.9  STATUS                         STARTED Compute graph
## 2020-05-30 10:32:09    39.7  STATUS                         COMPLETED Compute graph
## 2020-05-30 10:32:09    39.7  STATUS                         STARTED Compute clustering
## 2020-05-30 10:32:11    39.7  STATUS                         COMPLETED Compute clustering
## 2020-05-30 10:32:11    39.7  STATUS                     COMPLETED Compute correlation blocks
## 2020-05-30 10:32:11    39.7  STATUS                     STARTED Compute methQTL per correlation block
## 2020-05-30 10:32:11    39.7  STATUS                         STARTED Setting up Multicore
## 2020-05-30 10:32:11    39.7    INFO                             Using 1 cores
## 2020-05-30 10:32:11    39.7  STATUS                         COMPLETED Setting up Multicore
## 2020-05-30 11:05:43     6.1  STATUS                     COMPLETED Compute methQTL per correlation block
## 2020-05-30 11:05:43     6.1  STATUS                 COMPLETED Computing methQTL for chromosome chr17
## 2020-05-30 11:05:43     6.1  STATUS                 STARTED Computing methQTL for chromosome chr18
## 2020-05-30 11:05:43     6.1  STATUS                     STARTED Compute correlation blocks
## 2020-05-30 11:05:43     6.1  STATUS                         STARTED Compute correlation matrix
## 2020-05-30 11:05:47     7.2  STATUS                         COMPLETED Compute correlation matrix
## 2020-05-30 11:07:08    10.4  STATUS                         STARTED Compute pairwise distances
## 2020-05-30 11:07:13     6.1  STATUS                         COMPLETED Compute pairwise distances
## 2020-05-30 11:07:21     6.1  STATUS                         STARTED Weight distances
## 2020-05-30 11:08:06    13.3  STATUS                         COMPLETED Weight distances
## 2020-05-30 11:08:08     7.2  STATUS                         STARTED Compute graph
## 2020-05-30 11:08:12    10.9  STATUS                         COMPLETED Compute graph
## 2020-05-30 11:08:12    10.9  STATUS                         STARTED Compute clustering
## 2020-05-30 11:08:12    10.9  STATUS                         COMPLETED Compute clustering
## 2020-05-30 11:08:12    10.9  STATUS                     COMPLETED Compute correlation blocks
## 2020-05-30 11:08:13    10.9  STATUS                     STARTED Compute methQTL per correlation block
## 2020-05-30 11:08:13    10.9  STATUS                         STARTED Setting up Multicore
## 2020-05-30 11:08:13    10.9    INFO                             Using 1 cores
## 2020-05-30 11:08:13    10.9  STATUS                         COMPLETED Setting up Multicore
## 2020-05-30 11:34:53     6.1  STATUS                     COMPLETED Compute methQTL per correlation block
## 2020-05-30 11:34:53     6.1  STATUS                 COMPLETED Computing methQTL for chromosome chr18
## 2020-05-30 11:34:53     6.1  STATUS                 STARTED Computing methQTL for chromosome chr19
## 2020-05-30 11:34:53     6.1  STATUS                     STARTED Compute correlation blocks
## 2020-05-30 11:34:53     6.1  STATUS                         STARTED Compute correlation matrix
## 2020-05-30 11:35:22    19.9  STATUS                         COMPLETED Compute correlation matrix
## 2020-05-30 11:45:20    13.1  STATUS                         STARTED Compute pairwise distances
## 2020-05-30 11:45:40    11.8  STATUS                         COMPLETED Compute pairwise distances
## 2020-05-30 11:46:23     9.0  STATUS                         STARTED Weight distances
## 2020-05-30 11:51:22    20.3  STATUS                         COMPLETED Weight distances
## 2020-05-30 11:51:23    15.8  STATUS                         STARTED Compute graph
## 2020-05-30 11:52:09    29.5  STATUS                         COMPLETED Compute graph
## 2020-05-30 11:52:09    29.5  STATUS                         STARTED Compute clustering
## 2020-05-30 11:52:11    29.5  STATUS                         COMPLETED Compute clustering
## 2020-05-30 11:52:11    29.5  STATUS                     COMPLETED Compute correlation blocks
## 2020-05-30 11:52:12    29.5  STATUS                     STARTED Compute methQTL per correlation block
## 2020-05-30 11:52:12    29.5  STATUS                         STARTED Setting up Multicore
## 2020-05-30 11:52:12    29.5    INFO                             Using 1 cores
## 2020-05-30 11:52:12    29.5  STATUS                         COMPLETED Setting up Multicore
## 2020-05-30 12:14:59     6.1  STATUS                     COMPLETED Compute methQTL per correlation block
## 2020-05-30 12:14:59     6.1  STATUS                 COMPLETED Computing methQTL for chromosome chr19
## 2020-05-30 12:14:59     6.1  STATUS                 STARTED Computing methQTL for chromosome chr20
## 2020-05-30 12:14:59     6.1  STATUS                     STARTED Compute correlation blocks
## 2020-05-30 12:14:59     6.1  STATUS                         STARTED Compute correlation matrix
## 2020-05-30 12:15:09    11.4  STATUS                         COMPLETED Compute correlation matrix
## 2020-05-30 12:18:27    20.1  STATUS                         STARTED Compute pairwise distances
## 2020-05-30 12:18:36     8.7  STATUS                         COMPLETED Compute pairwise distances
## 2020-05-30 12:18:51     6.1  STATUS                         STARTED Weight distances
## 2020-05-30 12:20:41    14.0  STATUS                         COMPLETED Weight distances
## 2020-05-30 12:20:43     7.5  STATUS                         STARTED Compute graph
## 2020-05-30 12:20:56    12.7  STATUS                         COMPLETED Compute graph
## 2020-05-30 12:20:56    12.7  STATUS                         STARTED Compute clustering
## 2020-05-30 12:20:57    12.7  STATUS                         COMPLETED Compute clustering
## 2020-05-30 12:20:57    12.7  STATUS                     COMPLETED Compute correlation blocks
## 2020-05-30 12:20:57    12.7  STATUS                     STARTED Compute methQTL per correlation block
## 2020-05-30 12:20:57    12.7  STATUS                         STARTED Setting up Multicore
## 2020-05-30 12:20:57    12.7    INFO                             Using 1 cores
## 2020-05-30 12:20:57    12.7  STATUS                         COMPLETED Setting up Multicore
## 2020-05-30 12:47:22     6.1  STATUS                     COMPLETED Compute methQTL per correlation block
## 2020-05-30 12:47:22     6.1  STATUS                 COMPLETED Computing methQTL for chromosome chr20
## 2020-05-30 12:47:22     6.1  STATUS                 STARTED Computing methQTL for chromosome chr21
## 2020-05-30 12:47:22     6.1  STATUS                     STARTED Compute correlation blocks
## 2020-05-30 12:47:22     6.1  STATUS                         STARTED Compute correlation matrix
## 2020-05-30 12:47:24     6.6  STATUS                         COMPLETED Compute correlation matrix
## 2020-05-30 12:48:03     6.1  STATUS                         STARTED Compute pairwise distances
## 2020-05-30 12:48:06     6.1  STATUS                         COMPLETED Compute pairwise distances
## 2020-05-30 12:48:11     6.1  STATUS                         STARTED Weight distances
## 2020-05-30 12:48:23     6.6  STATUS                         COMPLETED Weight distances
## 2020-05-30 12:48:25     6.1  STATUS                         STARTED Compute graph
## 2020-05-30 12:48:25     6.1  STATUS                         COMPLETED Compute graph
## 2020-05-30 12:48:25     6.1  STATUS                         STARTED Compute clustering
## 2020-05-30 12:48:26     6.1  STATUS                         COMPLETED Compute clustering
## 2020-05-30 12:48:26     6.1  STATUS                     COMPLETED Compute correlation blocks
## 2020-05-30 12:48:26     6.1  STATUS                     STARTED Compute methQTL per correlation block
## 2020-05-30 12:48:26     6.1  STATUS                         STARTED Setting up Multicore
## 2020-05-30 12:48:26     6.1    INFO                             Using 1 cores
## 2020-05-30 12:48:26     6.1  STATUS                         COMPLETED Setting up Multicore
## 2020-05-30 12:48:26     6.1    INFO                         No SNP closer than 500000
## 2020-05-30 12:48:26     6.1    INFO                         No SNP closer than 500000
## 2020-05-30 12:48:27     6.1    INFO                         No SNP closer than 500000
## 2020-05-30 13:01:54     6.1  STATUS                     COMPLETED Compute methQTL per correlation block
## 2020-05-30 13:01:54     6.1  STATUS                 COMPLETED Computing methQTL for chromosome chr21
## 2020-05-30 13:01:54     6.1  STATUS                 STARTED Computing methQTL for chromosome chr22
## 2020-05-30 13:01:54     6.1  STATUS                     STARTED Compute correlation blocks
## 2020-05-30 13:01:54     6.1  STATUS                         STARTED Compute correlation matrix
## 2020-05-30 13:02:00     7.6  STATUS                         COMPLETED Compute correlation matrix
## 2020-05-30 13:04:16     6.1  STATUS                         STARTED Compute pairwise distances
## 2020-05-30 13:04:22     6.1  STATUS                         COMPLETED Compute pairwise distances
## 2020-05-30 13:04:35     6.1  STATUS                         STARTED Weight distances
## 2020-05-30 13:05:48     8.5  STATUS                         COMPLETED Weight distances
## 2020-05-30 13:05:49     7.6  STATUS                         STARTED Compute graph
## 2020-05-30 13:05:59    12.1  STATUS                         COMPLETED Compute graph
## 2020-05-30 13:05:59    12.1  STATUS                         STARTED Compute clustering
## 2020-05-30 13:06:00    12.1  STATUS                         COMPLETED Compute clustering
## 2020-05-30 13:06:00    12.1  STATUS                     COMPLETED Compute correlation blocks
## 2020-05-30 13:06:00    12.1  STATUS                     STARTED Compute methQTL per correlation block
## 2020-05-30 13:06:00    12.1  STATUS                         STARTED Setting up Multicore
## 2020-05-30 13:06:00    12.1    INFO                             Using 1 cores
## 2020-05-30 13:06:00    12.1  STATUS                         COMPLETED Setting up Multicore
## 2020-05-30 13:22:56     6.1  STATUS                     COMPLETED Compute methQTL per correlation block
## 2020-05-30 13:22:56     6.1  STATUS                 COMPLETED Computing methQTL for chromosome chr22
## 2020-05-30 13:23:05     7.0  STATUS             COMPLETED Computing methQTL for chromosome chr6</code></pre>
<p>We will now present the two steps of the methQTL calling procedure in more detail.</p>
<h2 id="compute-cpg-correlation-blocks">Compute CpG correlation blocks</h2>
<p>Since neighboring CpGs are often highly correlated, using each CpG independently as a potential methQTL candidate leads to many redundant results. We thus aimed to approximate <em>DNA methylation haplotypes</em> by determining highly correlated CpGs in close vicinity. The procedure itself is split into six steps, and is performed for each chromosome independently:</p>
<ol type="1">
<li>Compute the (Pearson) correlation matrix between all CpGs (futher correlation types available in option <code>correlation.type</code>)</li>
<li>Construct the distance matrix from the correlation matrix</li>
<li>Discard all interactions with a correlation lower than a given threshold (option: <code>cluster.cor.threshold</code>)</li>
<li>Weight the distance according to the genomic distance between the two CpGs with a Gaussian (option: <code>standard.deviation.gauss</code>). Higher values for the standard deviation lead to a lower penalty on distal CpGs, thus the cluster will become larger.</li>
<li>Discard all interactions ranging longer than the option <code>absolute.distance.cutoff</code></li>
<li>Compute the Louvain clustering on the undirected, weighted graph induced by the distance matrix</li>
</ol>
<p>This will return a clustering according to the correlation structure between neighboring CpGs that we will later use for methQTL calling. Note that we used simultation experiments to determine the parameters for each data type individually. They will be automatically loaded for the dataset that is used and are:</p>
<ul>
<li><strong>450k:</strong> <code>cluster.cor.threshold</code>=0.2, <code>standard.deviation.gauss</code>=5,000, <code>absolute.distance.cutoff</code>=500,000</li>
<li><strong>EPIC:</strong> <code>cluster.cor.threshold</code>=0.2, <code>standard.deviation.gauss</code>=3,000, <code>absolute.distance.cutoff</code>=500,000</li>
<li><strong>RRBS/WGBS:</strong> <code>cluster.cor.threshold</code>=0.2, <code>standard.deviation.gauss</code>=250, <code>absolute.distance.cutoff</code>=500,000</li>
</ul>
<h2 id="call-methqtl-per-correlation-block">Call methQTL per correlation block</h2>
<p>From the list of correlation blocks, <em>methQTL</em> computes methQTL interactions with all SNPs on the same chromosome. The process is split into three steps:</p>
<ol type="1">
<li>Compute a representative CpG per correlation block, as specified with the option <code>representative.cpg.computation</code> (default: <em>row.medians</em>).</li>
<li>Discard all SNPs that are further than <code>absolute.distance.cutoff</code> (default: 1,000,000) away from the representative CpG</li>
<li>Call methQTL by using linear models. Multiple options of methQTL calling are available and can be selected via the option <code>linear.model.type</code> (default: <em>classical.linear</em>). Alternatively, <em>fastQTL</em> can be set as an option for <code>meth.qtl.type</code>. This will tell the package to use the fastQTL software <span class="citation" data-cites="Ongen2016">(Ongen et al. 2016)</span>.</li>
</ol>
<p>The <code>meth.qtl.type</code> tells, how a methQTL interaction is defined and provides three options, in addition to the already mentioned <em>fastQTL</em>:</p>
<ol type="1">
<li><em>oneVSall</em>: A CpG can only be influenced by one SNP. We choose the one with the lowest p-value.</li>
<li><em>twoVSall</em>: A CpG can both positively and negatively be influenced by two independent SNPs. The package will output those fulfilling the p-value cutoff.</li>
<li><em>allVSall</em>: For each CpG, all SNPs showing a p-value lower than the p-value cutoff will be returned.</li>
</ol>
<p>In the latest stage, potential covariates can be specified using the option <em>sel.covariates</em>. We recommend to include at least <em>age</em> and <em>sex</em> as covariates, as they have a strong influence on the DNA methylation pattern.</p>
<h1 id="downstream-analysis-and-interpretation">Downstream analysis and interpretation</h1>
<h2 id="how-to-use-methqtlresult">How to use <em>methQTLResult</em></h2>
<p>The above procedure will create an object of class <code>methQTLResult</code>, which contains the methQTL that are called in the previous step. To get a table of all the methQTL, you need to extract the information from the object. In most of the function below, there is the option <code>type</code>, which takes on the values: * ‘SNP’: To characterize the SNPs that influence any DNA methylation state * ‘CpG’: To characterize the representative CpGs per correlation block that are influences by any genotype * ‘cor.block’: To characterize all CpGs, which are part of a correlation block, whose representative CpG is influenced by any genotype</p>
<p>Furthermore, you can obtain genomic annotations for both the CpGs and the SNPs involved in the methQTL interactions:</p>
<div class="sourceCode" id="cb15"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb15-1" data-line-number="1">result.table &lt;-<span class="st"> </span><span class="kw">getResult</span>(meth.qtl.res)</a>
<a class="sourceLine" id="cb15-2" data-line-number="2"><span class="kw">head</span>(result.table)</a></code></pre></div>
<pre><code>##            CpG        SNP       Beta      P.value Chromosome Position.CpG
## 91  cg00040738  rs3008409  0.9521075 1.128036e-06       chr1     31632794
## 354 cg00045114 rs10489293 -0.9731047 7.033229e-07       chr1    172674159
## 380 cg00090105  rs2038891  0.9311737 1.993551e-07       chr1    182538701
## 500 cg00102184  rs2045349 -1.4149612 2.694043e-07       chr1    240194783
## 476 cg00207189  rs3790973  0.6759185 7.521751e-07       chr1    231244795
## 153 cg00330609 rs10493281  0.6840097 3.069889e-06       chr1     60345404
##     Position.SNP Distance p.val.adj.fdr
## 91      32041838  -409044     1.0000000
## 354    172186787   487372     1.0000000
## 380    182367548   171153     0.6264101
## 500    240444305  -249522     0.7155086
## 476    230908863   335932     1.0000000
## 153     60739301  -393897     1.0000000</code></pre>
<div class="sourceCode" id="cb17"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb17-1" data-line-number="1">anno.meth &lt;-<span class="st"> </span><span class="kw">getAnno</span>(meth.qtl.res,<span class="st">&quot;meth&quot;</span>)</a>
<a class="sourceLine" id="cb17-2" data-line-number="2"><span class="kw">head</span>(anno.meth)</a></code></pre></div>
<pre><code>##            Chromosome  Start    End Strand Strand.1 AddressA AddressB Design
## cg26928153       chr1  10848  10849      -        - 91693541 47784201      I
## cg16269199       chr1  10850  10851      -        - 82663207  3701821      I
## cg24669183       chr1 534242 534243      -        - 12706847       NA     II
## cg15560884       chr1 710097 710098      +        + 66790119       NA     II
## cg01014490       chr1 714177 714178      -        -  1645492 53610280      I
## cg10692041       chr1 716245 716246      +        + 80722913       NA     II
##            Color Context Random HumanMethylation27 HumanMethylation450
## cg26928153   Grn      CG  FALSE              FALSE                  NA
## cg16269199   Grn      CG  FALSE              FALSE                  NA
## cg24669183  Both      CG  FALSE              FALSE                TRUE
## cg15560884  Both      CG  FALSE              FALSE                TRUE
## cg01014490   Red      CG  FALSE              FALSE                TRUE
## cg10692041  Both      CG  FALSE              FALSE                  NA
##            Mismatches.A Mismatches.B CGI.Relation CpG GC SNPs.3 SNPs.5
## cg26928153            0            0     Open Sea  15 74      0      0
## cg16269199            0            0     Open Sea  15 74      0      0
## cg24669183            0            0  South Shore   2 49      0      0
## cg15560884            0            0  North Shelf   2 28      0      0
## cg01014490            0            0       Island   9 67      0      0
## cg10692041            0            0  South Shore   2 47      0      0
##            SNPs.Full Cross.reactive
## cg26928153         0              0
## cg16269199         0              0
## cg24669183         0              0
## cg15560884         0              0
## cg01014490         0              0
## cg10692041         0              0</code></pre>
<div class="sourceCode" id="cb19"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb19-1" data-line-number="1">anno.geno &lt;-<span class="st"> </span><span class="kw">getAnno</span>(meth.qtl.res,<span class="st">&quot;geno&quot;</span>)</a>
<a class="sourceLine" id="cb19-2" data-line-number="2"><span class="kw">head</span>(anno.geno)</a></code></pre></div>
<pre><code>##            Chromosome  Start cM Allele.1 Allele.2 Allele.1.Freq Allele.2.Freq
## rs3094315        chr1 752566 NA        A        G        0.7500        0.2500
## rs3131972        chr1 752721 NA        G        A        0.7500        0.2500
## rs12124819       chr1 776546 NA        A        G        0.7500        0.2500
## rs11240777       chr1 798959 NA        G        A        0.7500        0.2500
## rs4970383        chr1 838555 NA        C        A        0.5625        0.4375
## rs4475691        chr1 846808 NA        C        T        0.6875        0.3125</code></pre>
<p>For more detailed information about the output, also see the function <code>getResults.GWASMap</code>.</p>
<h2 id="plots">Plots</h2>
<p>To visualize methQTL, the package provides some plotting functions. All functions return an object of type <code>ggplot</code>, which can be subsequently stored or viewed. Either all methQTL can be simultaneously visualized in a single plot, or a specific methQTL can be visualized:</p>
<div class="sourceCode" id="cb21"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb21-1" data-line-number="1">result.table &lt;-<span class="st"> </span>result.table[<span class="kw">order</span>(result.table<span class="op">$</span>P.value,<span class="dt">decreasing=</span>F),]</a>
<a class="sourceLine" id="cb21-2" data-line-number="2"><span class="kw">qtl.plot.SNP.CpG.interaction</span>(imp.data,result.table<span class="op">$</span>CpG[<span class="dv">1</span>],result.table<span class="op">$</span>SNP[<span class="dv">1</span>])</a></code></pre></div>
<figure>
<img src="figure/plots-1.png" alt="plot of chunk plots" /><figcaption>plot of chunk plots</figcaption>
</figure>
<div class="sourceCode" id="cb22"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb22-1" data-line-number="1"><span class="kw">qtl.distance.scatterplot</span>(meth.qtl.res)</a></code></pre></div>
<figure>
<img src="figure/plots-2.png" alt="plot of chunk plots" /><figcaption>plot of chunk plots</figcaption>
</figure>
<h2 id="interpretation-functions">Interpretation functions</h2>
<p>The package provides a bunch of interpretation functions to characterize the detected methQTLs. This includes LOLA enrichment analysis<span class="citation" data-cites="LOLA">(Sheffield and Bock 2016)</span> (<code>qtl.lola.enrichment</code>), genomic annotation enrichment based on putative regulatory elements defined by the Ensembl Regulatory Build<span class="citation" data-cites="Zerbino2015">(Zerbino et al. 2015)</span> (<code>qtl.annotation.enrichment</code>), enrichment analysis of different base substitutions in SNPs (<code>qtl.base.substitution.enrichment</code>), or TFBS motif enrichment using <a href="http://bioconductor.org/packages/release/bioc/html/TFBSTools.html">TFBSTools</a>. Enrichment is compared for the methQTLs that are available in the provided <code>methQTLResult</code> (for a single input), or to the overlapping QTLs for a list of `<code>methQTLResult</code>. The background of the enrichment is defined as all the SNPs/CpGs that have been used as input to the methQTL calling.</p>
<div class="sourceCode" id="cb23"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb23-1" data-line-number="1">res &lt;-<span class="st"> </span><span class="kw">qtl.base.substitution.enrichment</span>(meth.qtl.res)</a></code></pre></div>
<pre><code>## Warning in matrix(c(tps, fns, fps, tns), 2, 2): data length [3] is not a sub-
## multiple or multiple of the number of rows [2]</code></pre>
<div class="sourceCode" id="cb25"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb25-1" data-line-number="1"><span class="kw">qtl.plot.base.substitution</span>(meth.qtl.res,<span class="dt">merge=</span><span class="ot">TRUE</span>)</a></code></pre></div>
<h2 id="lists-of-methqtl-results">Lists of methQTL results</h2>
<p>Most of the functions discussed above either support a single <code>methQTLResult</code> as input, or a list of such objects. In case a list is specified, the functions with typically overlap the methQTLs found and compare those with all SNPs/CpGs that have been used for methQTL calling. Additionally, there are functions that particularly work on a list of <code>methQTLResult</code> objects and that perform overlapping, or determine the methQTLs specific to a dataset.</p>
<div class="sourceCode" id="cb26"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb26-1" data-line-number="1">meth.qtl.list &lt;-<span class="st"> </span><span class="kw">list</span>(<span class="dt">First=</span>meth.qtl.res<span class="fl">.1</span>,<span class="dt">Second=</span>meth.qtl.res<span class="fl">.2</span>,<span class="dt">Third=</span>meth.qtl.res<span class="fl">.3</span>)</a>
<a class="sourceLine" id="cb26-2" data-line-number="2"><span class="kw">qtl.venn.plot</span>(meth.qtl.list)</a>
<a class="sourceLine" id="cb26-3" data-line-number="3"><span class="kw">qtl.upset.plot</span>(meth.qtl.list,<span class="dt">type =</span> <span class="st">&quot;cor.block&quot;</span>)</a>
<a class="sourceLine" id="cb26-4" data-line-number="4">spec.first &lt;-<span class="st"> </span><span class="kw">get.specific.qtl</span>(meth.qtl.list<span class="op">$</span>First,meth.qtl.list[<span class="op">-</span><span class="dv">1</span>])</a></code></pre></div>
<h1 id="advanced-configuration">Advanced configuration</h1>
<h2 id="employ-methqtl-on-a-scientific-compute-cluster">Employ methQTL on a scientific compute cluster</h2>
<p><em>methQTL</em> can automatically distribute jobs across a high performance compute cluster, which has been setup using the Sun Grid Engine (SGE) technology. You can pass the option <code>cluster.submit</code> to <code>do.methQTL</code> and thus activate the cluster submission. Note that you’ll also have to specify a path to an executable <em>Rscript</em> and potentially specify resource requirements using the option setting <code>cluster.config</code>.</p>
<div class="sourceCode" id="cb27"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb27-1" data-line-number="1"><span class="kw">qtl.setOption</span>(<span class="dt">cluster.config =</span> <span class="kw">c</span>(<span class="dt">h_vmem=</span><span class="st">&quot;60G&quot;</span>,<span class="dt">mem_free=</span><span class="st">&quot;20G&quot;</span>))</a>
<a class="sourceLine" id="cb27-2" data-line-number="2"><span class="kw">qtl.setOption</span>(<span class="dt">rscript.path =</span> <span class="st">&quot;/usr/bin/Rscript&quot;</span>)</a>
<a class="sourceLine" id="cb27-3" data-line-number="3">meth.qtl.res &lt;-<span class="st"> </span><span class="kw">do.methQTL</span>(<span class="dt">meth.qtl =</span> imp.data,</a>
<a class="sourceLine" id="cb27-4" data-line-number="4">                          <span class="dt">cluster.submit =</span> T)</a></code></pre></div>
<div id="refs" class="references">
<div id="ref-Ongen2016">
<p>Ongen, Halit, Alfonso Buil, Andrew Anand Brown, Emmanouil T. Dermitzakis, and Olivier Delaneau. 2016. “Fast and efficient QTL mapper for thousands of molecular phenotypes.” <em>Bioinformatics</em> 32 (10): 1479–85. <a href="https://doi.org/10.1093/bioinformatics/btv722">https://doi.org/10.1093/bioinformatics/btv722</a>.</p>
</div>
<div id="ref-LOLA">
<p>Sheffield, Nathan C., and Christoph Bock. 2016. “LOLA: enrichment analysis for genomic region sets and regulatory elements in R and Bioconductor.” <em>Bioinformatics</em> 32 (4): 587–89. <a href="https://doi.org/10.1093/bioinformatics/btv612">https://doi.org/10.1093/bioinformatics/btv612</a>.</p>
</div>
<div id="ref-Sherry2001">
<p>Sherry, S. T., M.H. Ward, M. Kholodov, J. Baker, L. Phan, E.M. Smigielski, and K. Sirotkin. 2001. “dbSNP: the NCBI database of genetic variation.” <em>Nucleic Acids Research</em> 29 (1): 308–11. <a href="https://doi.org/10.1093/nar/29.1.308">https://doi.org/10.1093/nar/29.1.308</a>.</p>
</div>
<div id="ref-Zerbino2015">
<p>Zerbino, Daniel R., Steven P. Wilder, Nathan Johnson, Thomas Juettemann, and Paul R. Flicek. 2015. “The Ensembl Regulatory Build.” <em>Genome Biology</em> 16 (1). ??? 1–8. <a href="https://doi.org/10.1186/s13059-015-0621-5">https://doi.org/10.1186/s13059-015-0621-5</a>.</p>
</div>
</div>
